application/email/v0/read_emails.go

```diff
@@ -0,0 +1,274 @@
+		return emails, nil
+	}
+
+	if search.Limit == 0 {
+		search.Limit = EmailReadingDefaultCapacity
+	}
+	limit := search.Limit
+
+	// TODO: chuang8511, Research how to fetch emails by filter and concurrency.
+	// It will be done before 2024-07-26.
+	for i := selectedMbox.NumMessages; limit > 0; i-- {
+		limit--
+		email := Email{}
+		var seqSet imap.SeqSet
+		seqSet.AddNum(i)
+
+		fetchOptions := &imap.FetchOptions{
+			BodySection: []*imap.FetchItemBodySection{{}},
+		}
+		fetchCmd := c.Fetch(seqSet, fetchOptions)
+		msg := fetchCmd.Next()
+		var bodySection imapclient.FetchItemDataBodySection
+		var ok bool
+		for {
+			item := msg.Next()
+			if item == nil {
+				break
+			}
+			bodySection, ok = item.(imapclient.FetchItemDataBodySection)
+			if ok {
+				break
+			}
+		}
+		if !ok {
+			return nil, fmt.Errorf("FETCH command did not return body section")
+		}
+
+		mr, err := mail.CreateReader(bodySection.Literal)
+		if err != nil {
+			return nil, fmt.Errorf("FETCH command did not return body section")
+		}
+
+		h := mr.Header
+		setEnvelope(&email, h)
+
+		if !includeSearchCondition(email, search) {
+			if err := fetchCmd.Close(); err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		for {
+			p, err := mr.NextPart()
+			if err == io.EOF {
+				break
+			}
+			inlineHeader, ok := p.Header.(*mail.InlineHeader)
+			if ok && !isHTMLType(inlineHeader) {
+				b, _ := io.ReadAll(p.Body)
+				email.Message += string(b)
+			}
+		}
+
+		if !includeSearchMessage(email, search) {
+			if err := fetchCmd.Close(); err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		if err := fetchCmd.Close(); err != nil {
+			return nil, err
+		}
+
+		emails = append(emails, email)
+	}
+
+	return emails, nil
+}
+
+func convertMailboxString(mailbox string) string {
+	switch {
+	case mailbox == "INBOX":
+		return "INBOX"
+	case mailbox == "SENT":
+		return "[Gmail]/Sent Mail"
+	case mailbox == "DRAFTS":
+		return "[Gmail]/Drafts"
+	default:
+		return "INBOX"
+	}
+}
+
+func setEnvelope(email *Email, h mail.Header) {
+	if date, err := h.Date(); err != nil {
+		log.Fatalf("failed to get date: %v", err)
+	} else {
+		email.Date = date.Format(time.DateTime)
+	}
+	if from, err := h.AddressList("From"); err != nil {
+		log.Fatalf("failed to get from: %v", err)
+	} else {
+		email.From = from[0].String()
+	}
+	if to, err := h.AddressList("To"); err != nil {
+		log.Fatalf("failed to get to: %v", err)
+	} else {
+		email.To = []string{}
+		for _, t := range to {
+			email.To = append(email.To, t.String())
+		}
+	}
+	if subject, err := h.Text("Subject"); err != nil {
+		log.Fatalf("failed to get subject: %v", err)
+	} else {
+		email.Subject = subject
+	}
+}
+
+func includeSearchCondition(email Email, search Search) bool {
+	if search.SearchSubjectText != "" {
+		if !strings.Contains(email.Subject, search.SearchSubjectText) {
+			return false
+		}
+	}
+	if search.SearchFromEmail != "" {
+		if !strings.Contains(email.From, search.SearchFromEmail) {
```

Comment:
I suggest we perform an exact search here: `email.From == search.SearchFromEmail`. If the email address is `aaa@bbb.cc` but you search for `aa@bbb.cc`, things will go wrong.

Reply:
My original design is to let the users able to search `aaa@bbb.cc` through `aaa`.
But, I think your way will be better! So, I will modify it as well.

Reply:
Yeah, in this case, fuzzy search it not suitable.